### DISCUSSION TITLE
圣遗物评分OCR

### DISCUSSION THREAD

**Discussion by 490720818** [Arondight/Adachi-BOT#298] [Ideas]
参考项目[cocogoat](https://github.com/YuehaiTeam/cocogoat)

**Comment by Arondight** [comment]
如果需要本地再搭建一个 coco 的话，这个做个可配置的吧，我觉得默认走 api 比较好，因为部署起来很傻瓜。

如果自己仿写一个工作量太大了，我感觉为了 bot 这么搞有点不划算的

**Comment by Arondight** [comment]
我去看了看 coco 的代码，他这个硬编码的列表太草了。

https://github.com/YuehaiTeam/cocogoat/blob/main/src/ArtifactView/recognizeArtifact.ts

我上小学的时候（千年前）就有杂志说人工智能未来取代人类，这也太不靠谱了，认个字都固定老花眼，我对人类的未来感到无比的安心

**Comment by 490720818** [comment]
都是这个问题，特调那个api也是会把11.7%认成117%，11.1%认成1.1%。。。

**Comment by 490720818** [comment]
做成可修改肯定不合适，只能想办法提高识别率，有些识别错误可以优化掉的

**Comment by Arondight** [comment]
太强了，我最开始以为你要把 coco 放在 bot 里面，这下子跟着大佬后面学 ocr

**Comment by 490720818** [comment]
我之前用过cocogoat这个，错误率很低，实在不行写成api也行的，还有特调那个api看起来没人维护了，新本的圣遗物直接不支持

**Comment by Arondight** [comment]
写成什么无所谓的，就算直接可以用 coco 的二进制我觉得都没问题，部署起来不要麻烦就可以，如果让用户部署 bot 还要学会部署和配置 coco，那学习成本就太高了，当然能抄抄代码是最好的。

原作者直接用了 docker，不过我觉得还是手动更新对开发者来说舒服一些，现在 bot 也不稳定，三天两头就推个东西，尽量不要增加现在的部署步骤了。

**Comment by Arondight** [comment]
> 
> 
> 我之前用过cocogoat这个，错误率很低，实在不行写成api也行的，还有特调那个api看起来没人维护了，新本的圣遗物直接不支持

好像还真是，我评分一个充能毛，识别不了

---

我错了，是他识别的太不精确了，小数点给识别漏了，就出错了

**Comment by monsterxcn** [comment]
OCR 这个感觉可以用的工具还挺多的：QQ 自带识别文字（试了下对圣遗物识别不好用）、腾讯云 OCR（识别不出圣遗物星级，其他准确率感觉还可以）、百度 OCR、ocr.space，我是从另外一个 Node.js bot [@Tsuk1ko/cq-picsearcher-bot](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E9%99%84%E5%8A%A0%E5%8A%9F%E8%83%BD#ocr-%E6%96%87%E5%AD%97%E8%AF%86%E5%88%AB) 里看到的。

不怎么懂 Node.js，就用 Python 写了个大概的腾讯云识别然后处理的功能：[@monsterxcn/tencent_ocr_artifact.py](https://gist.github.com/monsterxcn/e67452e5ee6383495ff9bb0accc984fc) 

主要是不知道可莉特调圣遗物评分是怎么计算的，我看 API 返回的 `total_score` 是 `main_score` `sub_score` 之和，但是 `total_percent` 和 `main_percent` `sub_percent` 不知道是什么关系。看了看 NGA 社区的几篇文章，大概有这么几个套路：

1. [\[数据讨论\] \[提瓦特大学\]《圣遗物数值学导论》](https://ngabbs.com/read.php?tid=24270728)
   攻击力%：元素伤害%：暴击率%：暴击伤害%：物理伤害% = 1.5：1.5：1：2：1.875
2. [\[心得交流\] \[原神冒险团\] 圣遗物评分体系](https://ngabbs.com/read.php?tid=23802190)
   这个虽然说是副词条评分，但是给的表格里却有元素伤害加成、物理伤害加成、治疗加成，没看懂
   ```
   分数 = 攻击力% × 1.00 + 元素加成% × 1.00 + 生命值% × 1.00 + \
          物理加成 × 0.80 + 防御力% × 0.80 + 充能效率% × 0.90 + 治疗加成% × 1.30 + \
          暴击伤害% × 0.75 + 暴击率% × 1.50 + 元素精通 × 0.25 + \
          (攻击力%权重1.00 × 攻击力 / 基础攻击807) × 100 + \
          (生命值%权重1.00 × 生命值 / 基础生命5000) × 100 + \
          (防御力%权重0.80 × 防御力 / 基础防御500) × 100
   ```

第一个属性不全，至少元素精通、元素充能我感觉还是有点用的；第二个权重看起来好像有点奇怪（虽然我也不懂是怎么分析出来的。不知道还有没有别的评分公式了…

---------

顺手还翻到了一个计算圣遗物潜力的，公式我也没看懂，但是感觉圣遗物 OCR 做好了、再把这个公式整理下的话，这也可以做成一个功能：[\[心得交流\] 0717 ver. 全角色全玩法圣遗物胚子评分+角色定位一览](https://ngabbs.com/read.php?tid=26432895)

**Comment by Arondight** [comment]
@monsterxcn ok，谢谢这位老大哥
